Signal/src/environment/NotificationsManager.m

```diff
@@ -290,12 +302,47 @@ - (void)notifyUserForIncomingMessage:(TSIncomingMessage *)message
 
         [[PushManager sharedManager] presentNotification:notification checkForCancel:YES];
     } else {
-        if ([Environment.preferences soundInForeground]) {
+        if (shouldPlaySound && [Environment.preferences soundInForeground]) {
             AudioServicesPlayAlertSound(_newMessageSound);
         }
     }
 }
 
+- (BOOL)shouldPlaySoundForNotification
+{
+    OWSAssert([NSThread isMainThread]);
+
+    // Play no more than 2 notification sounds in a given
+    // five-second window.
+    const CGFloat kNotificationWindowSeconds = 5.f;
+    const NSUInteger kMaxNotificationRate = 2;
+
+    // Cull obsolete notification timestamps from the thread's notification history.
+    while (self.notificationHistory.count > 0) {
+        NSDate *notificationTimestamp = self.notificationHistory[0];
+        CGFloat notificationAgeSeconds = fabs(notificationTimestamp.timeIntervalSinceNow);
+        if (notificationAgeSeconds > kNotificationWindowSeconds) {
+            [self.notificationHistory removeObjectAtIndex:0];
```

Comment:
Since repeatedly removing the first element of an array is expensive, it's not ideal, but it seems like there's no issue because there will never be more than 2 items in this array. Right?